app/models/king.rb

```diff
@@ -9,8 +13,11 @@ def valid_move?(new_x, new_y)
   def can_move_out_of_check?
     ((x_position - 1)..(x_position + 1)).each do |x|
       ((y_position - 1)..(y_position + 1)).each do |y|
-        return true if valid_move?(x, y)
+        return true if valid_move?(x, y) &&
+                       game.under_attack?(is_white, x, y) == false &&
+                       (game.attacking_piece.type != 'knight' ? game.attacking_piece.straight_move?(x, y) == false : true)
```

Comment:
Since game.under_attack? already checks to see if an enemy piece can capture the piece at x, y, I don't understand why the last line (game.attacking_piece.type != 'knight' ? game.attacking_piece.straight_move?(x, y) == false : true) exits. Looking at valid move method for knight, it does not check for obstructions anyway, right?

Reply:
The last line exists because, if a king is in the line of attack via an attacking piece along a straight line (that's why I invoked straight_move), then the square which is on the _other_ side of the king from the attacking piece could be technically both a valid_move and not under_attack. But really, if the king moves there, then he will still be in check. So that square is not a real choice for him to move out of check. Does that make sense?

Reply:
So this still doesn't make sense to me. If an enemy piece can capture the king at that square where,  that should be still be accounted for at game.under_attack(is_white, x, y), which calls valid_move?(x, y) on the attacking piece. So it will check if there is rook at (1, 5) attacking the king at (3, 5), the valid_move method will still show the rook can capture the king if he moves to (4, 5).

Reply:
If there is a king at (3, 5), and a rook at (1, 5), then moving the rook to (4, 5) wouldn't be a valid move, because the king would be in the way.

However, if we check for a straight move, then we can also catch the instance where the king's current position blocks the rook's potential path. Does that help clear this up?

Reply:
Yes that makes sense. Interesting way to solve that case.